qpulseaudioengine.cpp

```diff
@@ -99,6 +100,26 @@ static void subscribeCallback(pa_context *context, pa_subscription_event_type_t
     }
 }
 
+static char quirk_primary_sink_name[PROP_VALUE_MAX];
+static char* quirk_sinkprimary_name(int& len)
+{
+    if ((len = property_get("t-o.quirk.forcesink", quirk_primary_sink_name, NULL)) > 0) {
```

Comment:
Could we not instead use gsettings for this, so that it's still usable without hybris?

Reply:
As this is a quirk for select devices I fail to see how to set gsettings properties from within the Halium port/adaptation.

Reply:
Sure, but we also still use telepathy-ofono on devices which do not have hybris. In the former days of Canonical images, the "custom" tarball included gsettings overrides, and is read-only in `/custom` path on the device. GSettings keys can also be "locked" so that generally the key could not be changed on those devices, as well. Unfortunately I think _pulseaudio_ doesn't use gsettings/dconf, so we can't take advantage of overrides and schema lock, without adding keys in our own schema for telepathy-ofono.

Reply:
Would those override files be required to change when we change a gsettings schema of telepathy-ofono? Just curious.
Also I think the Android properties are a leaner way of handling such quirks. Shouldn't telepathy-ofono work on non-hybris devices even with linking to libandroid-properties?

Reply:
The override files would only need to be changed if the schema was changed in some incompatible way. They aren't copies of schemas, but INI style files which specify what keys to override, and with what values, sort of like:
```
[org.compiz.gnomecompat]
main-menu-key=""
run-key=""
```
Even if telepathy-ofono would work without hybris while linking to libandroid-properties, it's something extra we have to include, which I think we can't use anyway. So, IMO, in general it is better to implement these things in some more generic way, so that we can bring the hybris and non-hybris builds even closer together, without just shipping useless things on the non-hybris builds.

Reply:
How about using https://github.com/ubports/deviceinfo for this? will gsetting work in the case? will it not needed to be compiled?

Reply:
We could use `deviceinfo` for this, sure. It would need to be rebuilt every time a change is added, so that new config ends up on devices, and it looks like it needs to be rebuilt for xenial already anyway. It wouldn't use GSettings in this case though, as it's a separate thing.

Reply:
Doing that will be faily simple,

just 
```
std::shared_ptr<DeviceInfo> info = std::make_shared<DeviceInfo>();

if (info->contains("AudioForceSink"){
  m_nametoset = info->get("AudioForceSink", "").c_str();
}

{same for source}

```

then put

then put  in a folder with the device name and place it in /etc/device-info
```
[device]
AudioForceSink=thesink
AudioForceSource=thesrc
```

Reply:
> We could use `deviceinfo` for this, sure. It would need to be rebuilt every time a change is added, so that new config ends up on devices, and it looks like it needs to be rebuilt for xenial already anyway. It wouldn't use GSettings in this case though, as it's a separate thing.

deviceinfo wont need to be recompiled to change a config? It needs to be included to xenial yes, but that should be fairly easy to do now that it's to a point where its ready.

Reply:
I've noticed that with failing getprop (due to running telepathy-ofono with the old AppArmor profile pre-reboot, just restarting the service) in-call-audio still works.
As it seems the only change we require on the Xperia X is setting the source/sink even if it matches the default.
We could idealy get rid of the quirk altogether as no other device has been reported to have working audio yet using it.

Reply:
Claims made before proper testing, seems the quirks are still needed. I'll take a look at deviceinfo. Should the ini file be provided as a halium override? @mariogrip

Reply:
> Claims made before proper testing, seems the quirks are still needed. I'll take a look at deviceinfo. Should the ini file be provided as a halium override? @mariogrip

If this is a device specific quirk then it will need to be a "device" override. for example if "hammerhead" needed this quirk i would add 
```
[device]
AudioForceSink=thesink
AudioForceSource=thesrc
```
to a file named /etc/device-info/hammerhead.conf

but if it's global to halium then it can be a halium override. I also plan to add another level of override, halium-version override.

Reply:
In short there is 4 levels of overrides:

device - priority 1
driver (halium or linux) - priority 2
type (desktop, phone, tablet) - priority 3
default - priority 4

Reply:
I've thought about it further and I personally think that we should proceed with the android property mechanism for the time being. The camera quirks by peat-psuwit for example also use the same mechanism and we shouldn't require porters to add configuration changes all over the place using different mechanisms. We're basically bike-shedding atm.